src/shims/fs.rs

```diff
@@ -257,6 +264,186 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         this.try_unwrap_io_result(result)
     }
 
+    fn statx(
+        &mut self,
+        dirfd_op: OpTy<'tcx, Tag>, // Should be an `int`
+        pathname_op: OpTy<'tcx, Tag>, // Should be a `const char *`
+        flags_op: OpTy<'tcx, Tag>, // Should be an `int`
+        _mask_op: OpTy<'tcx, Tag>, // Should be an `unsigned int`
+        statxbuf_op: OpTy<'tcx, Tag> // Should be a `struct statx *`
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("statx")?;
+
+        let statxbuf_scalar = this.read_scalar(statxbuf_op)?.not_undef()?;
+        let pathname_scalar = this.read_scalar(pathname_op)?.not_undef()?;
+
+        // If the statxbuf or pathname pointers are null, the function fails with `EFAULT`.
+        if this.is_null(statxbuf_scalar)? || this.is_null(pathname_scalar)? {
+            let efault = this.eval_libc("EFAULT")?;
+            this.set_last_error(efault)?;
+            return Ok(-1);
+        }
+
+        // Under normal circumstances, we would use `deref_operand(statxbuf_op)` to produce a
+        // proper `MemPlace` and then write the results of this function to it. However, the
+        // syscall API in Rust is untyped. This means that all the `statx` parameters are provided
+        // as `isize`s instead of having the proper types. Thus, we have to recover the layout of
+        // `statxbuf_op` by using the `libc::statx` struct type.
+        let statxbuf_place = {
+            // FIXME: This long path is required because `libc::statx` is an struct and also a
+            // function and `resolve_path` is returning the latter.
+            let statx_ty = this
+                .resolve_path(&["libc", "unix", "linux_like", "linux", "gnu", "statx"])?
+                .ty(*this.tcx);
+            let statxbuf_ty = this.tcx.mk_mut_ptr(statx_ty);
+            let statxbuf_layout = this.layout_of(statxbuf_ty)?;
+            let statxbuf_imm = ImmTy::from_scalar(statxbuf_scalar, statxbuf_layout);
+            this.ref_to_mplace(statxbuf_imm)?
+        };
+
+        let pathname: PathBuf = this.read_os_str_from_c_str(pathname_scalar)?.into();
+        // This argument should be a `c_int` but the syscall API just returns a pointer sized
+        // integer.
+        let flags: i32 = this.try_from_ptr_sized_operand(flags_op)?;
+
+        let at_fdcwd = this.eval_libc_i32("AT_FDCWD")?;
```

Comment:
Given that this is only needed inside the 2nd `if` branch, why compute it here already?

Reply:
I'm going to remove this whole chain of ifs and replace it for a single conditional checking if the path is valid.